quinine-core/src/main/scala/org/bdgenomics/quinine/metrics/rna/RNASeqPanelStats.scala

```diff
@@ -0,0 +1,174 @@
+ */
+object RNASeqPanelStats extends Serializable {
+
+  /**
+   * Generates aggregate count and coverage statistics for reads aligned to
+   * a transcriptome build.
+   *
+   * @param reads Reads aligned to a specific transcriptome build.
+   * @param transcriptome The transcriptome build, as loaded from a GTF/GFF or
+   *   generic feature file.
+   * @param startEndLength The amount of bases to consider as the 5'/3' end of
+   *   the transcript.
+   * @return Returns the aggregate stats.
+   */
+  def apply(reads: RDD[AlignmentRecord],
+            transcriptome: RDD[Feature],
+            startEndLength: Int): RNASeqPanelStats = {
+
+    // cache the transcriptome build
+    transcriptome.cache()
+
+    // cache the reads
+    reads.cache()
+
+    // compute the raw read alignment stats
+    val counts = RNASeqReadCounts(reads)
+
+    // join the reads against the transcriptome description
+    val joined = RightOuterJoin(transcriptome.keyBy(ReferenceRegion(_)),
+      reads.filter(_.getReadMapped)
+        .keyBy(ReferenceRegion(_)))
+      .cache()
+
+    // discard the alignments
+    val featureOptRdd = joined.map(kv => kv._1)
+
+    // run the alignment stat aggregator
+    val alignmentStats = RNASeqAlignmentCounts(featureOptRdd)
+
+    // run the transcript count bit
+    val transcriptStats = TranscriptCounts(transcriptome, featureOptRdd)
+
+    // compute coverage stats
+    val coverageObservations = ReadCoverage(reads)
+      .cache()
+
+    // unpersist reads
+    reads.unpersist()
+
+    // join coverage data against transcript features
+    val targetCoverageSites = BroadcastRegionJoin.partitionAndJoin(transcriptome.filter(_.getFeatureType == "transcript")
+      .keyBy(ReferenceRegion(_)),
+      coverageObservations.map(kv => {
+        val (pos, obs) = kv
+        (ReferenceRegion(pos), (pos, obs))
+      }))
+      .cache()
+
+    // compute mean coverage per transcript
+    val coverageStatsAllBases = CoverageStats(targetCoverageSites.map(kv => {
+      val (_, (site, coverage)) = kv
+      (site, coverage)
+    }))
+
+    // splits out the first/last _n_ bases of a reference region
+    def fragment(rr: ReferenceRegion): Seq[ReferenceRegion] = {
+      if (rr.length <= startEndLength) {
+        Seq(rr)
+      } else {
+        Seq(ReferenceRegion(rr.referenceName, rr.start, rr.start + startEndLength),
+          ReferenceRegion(rr.referenceName, rr.end - startEndLength, rr.end))
+      }
+    }
+
+    // compute mean coverage at the ends of transcripts
+    val coverageStatsStartEnd = CoverageStats(targetCoverageSites.flatMap(kv => {
+      val (region, (site, coverage)) = kv
+
+      // demorgan's law up in this
+      // we keep the site if it overlaps with at least one of the start/end of a region
+      // or, in other words, if it doesn't not overlap with all regions
+      if (fragment(ReferenceRegion(region)).forall(r => !r.overlaps(site))) {
+        None
+      } else {
+        Some((site, coverage))
```

Comment:
Basically, you only return `(site, coverage)` pairs that overlap exactly one end of the fragment (or just the region itself, if it's shorter than the minimum length), is that correct?